worker/raft/raftleaseconsumer/handler.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2021 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+package raftleaseconsumer
+
+import (
+	"net/http"
+	"time"
+
+	gorillaws "github.com/gorilla/websocket"
+	"github.com/juju/clock"
+
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/apiserver/websocket"
+)
+
+// Handler is an http.Handler suitable for serving lease consuming connections.
+type Handler struct {
+	operations chan<- operation
+	abort      <-chan struct{}
+	clock      clock.Clock
+	logger     Logger
+}
+
+// NewHandler returns a new Handler that sends operations to the
+// given operations channel, and stops accepting operations after
+// the abort channel is closed.
+func NewHandler(
+	operations chan<- operation,
+	abort <-chan struct{},
+	clock clock.Clock,
+	logger Logger,
+) *Handler {
+	return &Handler{
+		operations: operations,
+		abort:      abort,
+		clock:      clock,
+		logger:     logger,
+	}
+}
+
+// ServeHTTP is part of the http.Handler interface.
+func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	handler := func(socket *websocket.Conn) {
+		h.logger.Debugf("start of *raftleaseconsumer.ServeHTTP")
+		defer socket.Close()
+
+		// If we get to here, no more errors to report, so we report a nil
+		// error.  This way the first line of the socket is always a json
+		// formatted simple error.
+		h.sendError(socket, nil)
+
+		// Ensure we set the read deadline on the sockets for the Ping/Pong
+		// messages. That way we can workout if a client has actually gone away
+		// or not.
+		h.ensureReadDeadline(socket)
+
+		ticker := time.NewTicker(websocket.PingPeriod)
+		defer ticker.Stop()
+
+		operations := h.receiveOperations(socket)
+		for {
+			select {
+			case <-h.abort:
+				return
+			case <-ticker.C:
+				deadline := time.Now().Add(websocket.WriteWait)
+				if err := socket.WriteControl(gorillaws.PingMessage, []byte{}, deadline); err != nil {
+					// This error is expected if the other end goes away. By
+					// returning we close the socket through the defer call.
+					h.logger.Debugf("failed to write ping: %s", err)
+					return
+				}
+			case op := <-operations:
+				select {
+				case <-h.abort:
+					return
+				case h.operations <- op:
+				}
+			}
+		}
+	}
+	websocket.Serve(w, r, handler)
+}
+
+func (h *Handler) receiveOperations(socket *websocket.Conn) <-chan operation {
+	operations := make(chan operation)
+
+	go func() {
+		for {
+			// The message needs to be new each time through the loop to ensure
+			// the map is not reused.
+			var m params.LeaseOperation
+			// Receive() blocks until data arrives but will also be
+			// unblocked when the API handler calls socket.Close as it
+			// finishes.
+			if err := socket.ReadJSON(&m); err != nil {
+				h.handleSocketError(err)
+				return
```

Comment:
Do we need to do more than just return here? 

It looks like we'll just stop processing messages, but we haven't exited the higher level handler.